File: TP1-parte3/02_Python/scripts/verificar.py
"""Pruebas de estrés del cruce de babuinos.

Comprueba dos cosas:
1. La cuerda real nunca deja que se crucen, la rompan o entren fuera de turno, aun con
   muchos babuinos y tiempos muy comprimidos (mucha más contención que la corrida normal).
2. El auditor de verdad detecta los errores: se le pasan dos cuerdas defectuosas a propósito
   (sin control de dirección y sin límite de capacidad) y tiene que dar la alarma.
"""
import pathlib
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent.parent / "src"))

from constantes import CAPACIDAD_CUERDA  # noqa: E402
from cuerda import Cuerda  # noqa: E402
from simulacion import simular  # noqa: E402

ESCALA_RAPIDA = 0.05
REPETICIONES = 10
INTENTOS_DETECCION = 10
CONFIGURACIONES = [(1, 0), (0, 1), (5, 5), (8, 6), (0, 40), (40, 0), (25, 25), (60, 60)]
CODIGO_ERROR = 1


class CuerdaSinControlDeDireccion(Cuerda):
    """Defectuosa: deja subir a cualquiera mientras haya lugar (los babuinos se cruzan)."""

    def _puede_subir(self, turno, direccion):
        return turno == self._turno_que_puede_subir and self._en_cuerda < CAPACIDAD_CUERDA


class CuerdaSinLimiteDeCapacidad(Cuerda):
    """Defectuosa: no cuenta cuántos hay en la cuerda (se rompe)."""

    def _puede_subir(self, turno, direccion):
        return turno == self._turno_que_puede_subir and (self._en_cuerda == 0 or self._direccion is direccion)


def verificar_configuracion(izquierda, derecha):
    """Corre la configuración varias veces. Devuelve (todo_ok, maximo_simultaneo_visto)."""
    todo_ok, maximo = True, 0
    for _ in range(REPETICIONES):
        informe = simular(izquierda, derecha, escala_tiempo=ESCALA_RAPIDA)
        maximo = max(maximo, informe.maximo_simultaneo)
        todo_ok = todo_ok and informe.es_correcto() and informe.total_cruzaron == izquierda + derecha
    return todo_ok, maximo


def el_auditor_detecta(clase_cuerda, izquierda, derecha, campo_de_error):
    """True si, en algún intento, el auditor marca el error esperado con la cuerda defectuosa."""
    for _ in range(INTENTOS_DETECCION):
        informe = simular(izquierda, derecha, escala_tiempo=ESCALA_RAPIDA, clase_cuerda=clase_cuerda)
        if getattr(informe, campo_de_error) > 0:
            return True
    return False


def verificar_cuerda_real():
    todo_ok, maximo_global = True, 0
    for izquierda, derecha in CONFIGURACIONES:
        ok, maximo = verificar_configuracion(izquierda, derecha)
        maximo_global = max(maximo_global, maximo)
        todo_ok = todo_ok and ok
        print(f"  {izquierda:>2} izquierda / {derecha:>2} derecha x{REPETICIONES}: {'OK' if ok else 'FALLA'} (máximo en cuerda: {maximo})")
    alcanzo_capacidad = maximo_global == CAPACIDAD_CUERDA
    print(f"  Se llegó a la capacidad máxima ({CAPACIDAD_CUERDA}) sin pasarse: {'sí' if alcanzo_capacidad else 'NO'}")
    return todo_ok and alcanzo_capacidad


def verificar_auditor():
    detecta_peleas = el_auditor_detecta(CuerdaSinControlDeDireccion, 20, 20, "peleas")
    detecta_roturas = el_auditor_detecta(CuerdaSinLimiteDeCapacidad, 0, 60, "cuerdas_rotas")
    print(f"  Cuerda sin control de dirección -> el auditor detecta peleas: {'sí' if detecta_peleas else 'NO'}")
    print(f"  Cuerda sin límite de capacidad  -> el auditor detecta roturas: {'sí' if detecta_roturas else 'NO'}")
    return detecta_peleas and detecta_roturas


def main():
    sys.stdout.reconfigure(encoding="utf-8")
    print("1) Cuerda real bajo estrés")
    cuerda_ok = verificar_cuerda_real()
    print("2) El auditor detecta cuerdas defectuosas")
    auditor_ok = verificar_auditor()
    print("\nRESULTADO:", "TODO OK" if cuerda_ok and auditor_ok else "HAY FALLAS")
    return 0 if cuerda_ok and auditor_ok else CODIGO_ERROR


if __name__ == "__main__":
    sys.exit(main())

File: TP1-parte3/01_C++/tests/verificar.cpp
// Pruebas de estres del bano compartido.
//
// Comprueba dos cosas:
// 1. El bano real nunca mezcla hombres y mujeres, nunca supera la capacidad y respeta el turno, aun con
//    muchos empleados y tiempos muy comprimidos (mucha mas contencion que la corrida normal).
// 2. El auditor de verdad detecta los errores: se le pasan dos politicas defectuosas a proposito (sin
//    control de sexo y sin limite de capacidad) y tiene que dar la alarma.

#include <algorithm>
#include <iostream>
#include <vector>

#include "../src/constantes.h"
#include "../src/observador.h"
#include "../src/politica_admision.h"
#include "../src/simulacion.h"

namespace
{

  constexpr double kEscalaRapida = 0.05;
  constexpr int kRepeticiones = 10;
  constexpr int kIntentosDeteccion = 10;

  struct Configuracion
  {
    int hombres;
    int mujeres;
  };

  const std::vector<Configuracion> kConfiguraciones = {{1, 0}, {0, 1}, {5, 5}, {8, 6}, {0, 40}, {40, 0}, {25, 25}, {60, 60}};

  // Defectuosa: deja entrar a cualquiera mientras haya lugar (hombres y mujeres se mezclan).
  class PoliticaSinControlDeSexo : public bano::IPoliticaAdmision
  {
   public:
    bool PuedeEntrar(const bano::EstadoBano& estado, int turno, bano::Sexo) const override
    {
      return turno == estado.turno_que_puede_entrar && estado.adentro < bano::kCapacidadBano;
    }
  };

  // Defectuosa: no cuenta cuantos hay adentro (se supera la capacidad).
  class PoliticaSinLimiteDeCapacidad : public bano::IPoliticaAdmision
  {
   public:
    bool PuedeEntrar(const bano::EstadoBano& estado, int turno, bano::Sexo sexo) const override
    {
      return turno == estado.turno_que_puede_entrar && (estado.adentro == 0 || estado.sexo_actual == sexo);
    }
  };

  bool VerificarConfiguracion(const Configuracion& configuracion, int& maximo_visto)
  {
    const bano::PoliticaFifoEstricta politica;
    bool todo_ok = true;
    for (int repeticion = 0; repeticion < kRepeticiones; ++repeticion)
    {
      const bano::InformeAuditoria informe = bano::Simular(configuracion.hombres, configuracion.mujeres, politica, {}, kEscalaRapida);
      maximo_visto = std::max(maximo_visto, informe.maximo_simultaneo);
      todo_ok = todo_ok && informe.EsCorrecto() && informe.TotalQueUsaron() == configuracion.hombres + configuracion.mujeres;
    }
    return todo_ok;
  }

  bool VerificarBanoReal()
  {
    bool todo_ok = true;
    int maximo_global = 0;
    for (const Configuracion& configuracion : kConfiguraciones)
    {
      int maximo = 0;
      const bool ok = VerificarConfiguracion(configuracion, maximo);
      maximo_global = std::max(maximo_global, maximo);
      todo_ok = todo_ok && ok;
      std::cout << "  " << configuracion.hombres << " hombres / " << configuracion.mujeres << " mujeres x" << kRepeticiones << ": " << (ok ? "OK" : "FALLA")
                << " (maximo adentro: " << maximo << ")\n";
    }
    const bool alcanzo_capacidad = maximo_global == bano::kCapacidadBano;
    std::cout << "  Se llego a la capacidad maxima (" << bano::kCapacidadBano << ") sin pasarse: " << (alcanzo_capacidad ? "si" : "NO") << "\n";
    return todo_ok && alcanzo_capacidad;
  }

  // True si, en algun intento, el auditor marca el error esperado con la politica defectuosa.
  template <typename Politica, typename Campo>
  bool ElAuditorDetecta(int hombres, int mujeres, Campo campo)
  {
    const Politica politica;
    for (int intento = 0; intento < kIntentosDeteccion; ++intento)
    {
      const bano::InformeAuditoria informe = bano::Simular(hombres, mujeres, politica, {}, kEscalaRapida);
      if (informe.*campo > 0)
      {
        return true;
      }
    }
    return false;
  }

  bool VerificarAuditor()
  {
    const bool detecta_mezclas = ElAuditorDetecta<PoliticaSinControlDeSexo>(20, 20, &bano::InformeAuditoria::mezclas);
    const bool detecta_excesos = ElAuditorDetecta<PoliticaSinLimiteDeCapacidad>(0, 60, &bano::InformeAuditoria::excesos_de_capacidad);
    std::cout << "  Politica sin control de sexo   -> el auditor detecta mezclas: " << (detecta_mezclas ? "si" : "NO") << "\n";
    std::cout << "  Politica sin limite de capacidad -> el auditor detecta excesos: " << (detecta_excesos ? "si" : "NO") << "\n";
    return detecta_mezclas && detecta_excesos;
  }

}  // namespace

int main()
{
  std::cout << "1) Bano real bajo estres\n";
  const bool bano_ok = VerificarBanoReal();
  std::cout << "2) El auditor detecta politicas defectuosas\n";
  const bool auditor_ok = VerificarAuditor();
  const bool todo_ok = bano_ok && auditor_ok;
  std::cout << "\nRESULTADO: " << (todo_ok ? "TODO OK" : "HAY FALLAS") << std::endl;
  return todo_ok ? 0 : bano::kCodigoError;
}
